File: fixtures/lifecycle-evidence-and-record/verify.py
#!/usr/bin/env python3
# Copyright 2026 Tymofii Pidlisnyi. Apache-2.0 license. See LICENSE.
#
# Independent Python runner for the lifecycle-evidence-and-record candidate
# family.
#
# Written from README.md and vectors.json rather than ported from harness.ts.
# The retention clock here is computed with datetime.replace on the calendar
# year and the receipt group is a dispatch table keyed on (query, later
# evidence kind), neither of which mirrors the TypeScript file's shape, so the
# two implementations agreeing is evidence about the rules.
#
# The only third-party import is the published PyPI agent-passport-system, for
# its RFC 8785 (JCS) canonicalization primitive.
#
# Run:
#   /path/to/venv/bin/python fixtures/lifecycle-evidence-and-record/verify.py
#
# Exit 0 on success, 1 on any failure.

from __future__ import annotations

import hashlib
import json
import sys
from datetime import datetime, timezone
from pathlib import Path

from agent_passport.canonical import canonicalize_jcs

HERE = Path(__file__).resolve().parent
VOCABULARY = {"valid", "invalid", "not_established", "not_yet_effective", "suspended", "restricted"}

problems: list[str] = []


def fail(message: str) -> None:
    problems.append(message)
    print(f"  FAIL {message}", file=sys.stderr)


def at(iso: str) -> datetime:
    return datetime.fromisoformat(iso.replace("Z", "+00:00")).astimezone(timezone.utc)


# --------------------------------------------------------------------------
# retention_restriction
# --------------------------------------------------------------------------


def retention(inp: dict, *, own_clock: bool) -> tuple[str, str, str]:
    chain = inp["requester_chain_state"]
    record = inp["record"]

    if inp["requested_effect"] != "records:delete":
        return chain, "valid", "evidence_survives_authority_end"

    if record["trigger_occurred_at"] is None:
        return chain, "restricted", "retention_trigger_not_reached"

    if not own_clock:
        # The relationship ended, so a relationship-scoped duty is over.
        return chain, "valid", "retention_duty_elapsed"

    trigger = at(record["trigger_occurred_at"])
    # A duty stated in years advances the calendar year rather than adding a
    # fixed span, so it does not drift by a day per leap year. No date in this
    # family is 29 February.
    expiry = trigger.replace(year=trigger.year + record["retention_years"])
    if at(inp["evaluated_at"]) < expiry:
        return chain, "restricted", "retention_duty_unexpired"

    for hold in inp["external_holds"]:
        if not hold["released"]:
            return chain, "restricted", "external_hold_active"

    return chain, "valid", "retention_duty_elapsed"


# --------------------------------------------------------------------------
# receipt_immutability
# --------------------------------------------------------------------------

PRESENT_TENSE = {
    "new_information": ("invalid", "invalid", "new_record_references_prior"),
    "void_from_inception": ("invalid", "invalid", "ancestor_never_validly_issued"),
}


def receipt(inp: dict, *, honour_decision_time: bool) -> tuple[str, str, str]:
    prior = inp["prior_receipt"]
    kind = inp["later_evidence"]["kind"]

    if honour_decision_time:
        if prior["actual_bytes_sha256"] != prior["recorded_bytes_sha256"]:
            return "not_established", "not_established", "prior_receipt_bytes_do_not_match_digest"
        if inp["query"] == "verdict_at_decision_time":
            recorded = prior["recorded_verdict"]
            return recorded, recorded, "decided_on_the_record_then_available"
        if kind in PRESENT_TENSE:
            return PRESENT_TENSE[kind]
        recorded = prior["recorded_verdict"]
        return recorded, recorded, "no_later_record"

    # N2: answer everything against what is known now, never check the bytes.
    if kind in PRESENT_TENSE:
        return PRESENT_TENSE[kind]
    recorded = prior["recorded_verdict"]
    return recorded, recorded, "decided_on_the_record_then_available"


REFERENCE = {
    "retention_restriction": lambda i: retention(i, own_clock=True),
    "receipt_immutability": lambda i: receipt(i, honour_decision_time=True),
}

CONTROL = {
    "retention_restriction": lambda i: retention(i, own_clock=False),
    "receipt_immutability": lambda i: receipt(i, honour_decision_time=False),
}


def expected_triple(case: dict) -> tuple[str, str, str]:
    e = case["expected"]
    return e["chain_verdict"], e["verdict"], e["reason"]


def main() -> int:
    vectors = json.loads((HERE / "vectors.json").read_text())
    pinned = json.loads((HERE / "input-digests.json").read_text())["inputs"]
    cases = vectors["cases"]

    print(f"{vectors['family']}: {len(cases)} cases, Python runner")

    if vectors["status"] != "candidate_against_proposed":
        fail(f"family status is {vectors['status']!r}")
    if not vectors.get("proposed_text", {}).get("commit"):
        fail("family does not name the proposed text it is a candidate against")

    ids = set()
    for case in cases:
        if case["id"] in ids:
            fail(f"{case['id']}: duplicate vector id")
        ids.add(case["id"])
        if case["label"] != "candidate_against_proposed":
            fail(f"{case['id']}: label is {case['label']!r}")
        if not case["id"].startswith(case["case_id"]):
            fail(f"{case['id']}: vector id does not carry its CASES.md case id")
        if not case.get("proposed_text", {}).get("sections"):
            fail(f"{case['id']}: names no proposed text")
        for field in ("chain_verdict", "verdict"):
            if case["expected"][field] not in VOCABULARY:
                fail(f"{case['id']}: {field} {case['expected'][field]!r} is outside the settled vocabulary")
    print(f"  ok   labelling: {len(cases)} cases carry candidate_against_proposed and name their proposed text")

    before = {}
    for case in cases:
        digest = hashlib.sha256(canonicalize_jcs(case["input"]).encode("utf-8")).hexdigest()
        before[case["id"]] = digest
        if pinned.get(case["id"]) is None:
            fail(f"{case['id']}: no pinned input digest")
        elif digest != pinned[case["id"]]:
            fail(f"{case['id']}: input digest {digest} does not match pinned {pinned[case['id']]}")
    if len(pinned) != len(cases):
        fail(f"input-digests.json pins {len(pinned)} inputs for {len(cases)} cases")
    print(f"  ok   canonical bytes: {len(cases)}/{len(cases)} JCS input digests match, via PyPI canonicalize_jcs")

    matched = 0
    for case in cases:
        got = REFERENCE[case["group"]](case["input"])
        want = expected_triple(case)
        if got != want:
            fail(f"{case['id']}: reference gave {'/'.join(got)}, expected {'/'.join(want)}")
        else:
            matched += 1
    print(f"  ok   reference model: {matched}/{len(cases)} matched")

    for control in vectors["policies"]["negative_controls"]:
        group = control["group"]
        in_group = [c for c in cases if c["group"] == group]
        observed = [c["id"] for c in in_group if CONTROL[group](c["input"]) != expected_triple(c)]
        declared = sorted(control["declared_fail_set"])
        if sorted(observed) != declared:
            fail(f"{control['id']}: declared fail set {declared}, observed {sorted(observed)}")
        else:
            print(
                f"  ok   {control['id']}: ran {len(in_group)} cases in group {group}, "
                f"failed exactly {len(declared)} declared ({', '.join(declared)})"
            )
        if not declared:
            fail(f"{control['id']}: a negative control with an empty declared fail set is not a control")

    unchanged = 0
    for case in cases:
        after = hashlib.sha256(canonicalize_jcs(case["input"]).encode("utf-8")).hexdigest()
        if after != before[case["id"]]:
            fail(f"{case['id']}: an evaluator mutated the input it was given")
        else:
            unchanged += 1
    print(f"  ok   input immutability: {unchanged}/{len(cases)} inputs unchanged after every evaluator ran")

    if problems:
        print(f"\nlifecycle-evidence-and-record Python: FAILED, {len(problems)} problem(s)", file=sys.stderr)
        return 1
    print(f"\nlifecycle-evidence-and-record Python: {len(cases)}/{len(cases)} passed")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: fixtures/lifecycle-evidence-and-record/verify.ts
// Copyright 2026 Tymofii Pidlisnyi. Apache-2.0 license. See LICENSE.
//
// TypeScript runner for the lifecycle-evidence-and-record candidate family.
//
// Four checks, all of which must pass:
//
//   1. Labelling. Every case carries status candidate_against_proposed, names
//      the proposed text it tests, carries its CASES.md case id in its vector
//      id, and uses only the settled verdict vocabulary for both of its two
//      verdict fields.
//   2. Canonical bytes. Every case's `input` recanonicalizes under RFC 8785 to
//      its pinned SHA-256, through the published npm package's canonicalizeJCS.
//   3. Reference model. Each group's reference evaluator matches the case's
//      expected chain_verdict, verdict and reason.
//   4. Negative controls. Each group's control runs against every case in that
//      group, and its observed fail set must equal its declared fail set.
//
//   5. Input immutability. The receipt_immutability group is about a record not
//      being rewritten, so this runner also checks that no evaluator mutated
//      the input it was handed, by recomputing every input digest a second time
//      after all evaluators have run. A harness that edited a receipt in place
//      to produce its answer would pass check 3 and fail here.
//
// Run: npx tsx fixtures/lifecycle-evidence-and-record/verify.ts
// Exit 0 on success, 1 on any failure.

import { readFileSync } from 'node:fs'
import { createHash } from 'node:crypto'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'

import { canonicalizeJCS } from 'agent-passport-system'

import { CONTROL_FOR_GROUP, GROUPS, evaluateControl, evaluateReference, type Group, type Outcome } from './harness.ts'

const HERE = dirname(fileURLToPath(import.meta.url))
const VOCABULARY = new Set(['valid', 'invalid', 'not_established', 'not_yet_effective', 'suspended', 'restricted'])

interface Case {
  id: string
  case_id: string
  group: Group
  role: string
  label: string
  proposed_text: { sections: string[] }
  input: unknown
  expected: Outcome
}

const vectors = JSON.parse(readFileSync(join(HERE, 'vectors.json'), 'utf8')) as {
  family: string
  status: string
  proposed_text: { repository: string; commit: string }
  policies: { negative_controls: { id: string; group: Group; declared_fail_set: string[] }[] }
  cases: Case[]
}
const pinned: Record<string, string> = JSON.parse(readFileSync(join(HERE, 'input-digests.json'), 'utf8')).inputs

const failures: string[] = []
const fail = (m: string) => {
  failures.push(m)
  console.error(`  FAIL ${m}`)
}

const digestOf = (input: unknown) =>
  createHash('sha256').update(Buffer.from(canonicalizeJCS(input as never), 'utf8')).digest('hex')

console.log(`${vectors.family}: ${vectors.cases.length} cases, TypeScript runner`)

// --- 1. Labelling -----------------------------------------------------------

if (vectors.status !== 'candidate_against_proposed') fail(`family status is "${vectors.status}"`)
if (!vectors.proposed_text?.repository || !vectors.proposed_text?.commit) {
  fail('family does not name the proposed text it is a candidate against')
}
const seen = new Set<string>()
for (const c of vectors.cases) {
  if (seen.has(c.id)) fail(`${c.id}: duplicate vector id`)
  seen.add(c.id)
  if (c.label !== 'candidate_against_proposed') fail(`${c.id}: label is "${c.label}"`)
  if (!c.case_id || !c.id.startsWith(c.case_id)) fail(`${c.id}: vector id does not carry its CASES.md case id`)
  if (!Array.isArray(c.proposed_text?.sections) || c.proposed_text.sections.length === 0) {
    fail(`${c.id}: names no proposed text`)
  }
  if (!GROUPS.includes(c.group)) fail(`${c.id}: unknown group "${c.group}"`)
  for (const field of ['chain_verdict', 'verdict'] as const) {
    if (!VOCABULARY.has(c.expected[field])) {
      fail(`${c.id}: ${field} "${c.expected[field]}" is outside the settled vocabulary`)
    }
  }
}
console.log(`  ok   labelling: ${vectors.cases.length} cases carry candidate_against_proposed and name their proposed text`)

// --- 2. Canonical bytes -----------------------------------------------------

const before = new Map<string, string>()
for (const c of vectors.cases) {
  const got = digestOf(c.input)
  before.set(c.id, got)
  if (pinned[c.id] === undefined) fail(`${c.id}: no pinned input digest`)
  else if (got !== pinned[c.id]) fail(`${c.id}: input digest ${got} does not match pinned ${pinned[c.id]}`)
}
if (Object.keys(pinned).length !== vectors.cases.length) {
  fail(`input-digests.json pins ${Object.keys(pinned).length} inputs for ${vectors.cases.length} cases`)
}
console.log(`  ok   canonical bytes: ${vectors.cases.length}/${vectors.cases.length} JCS input digests match, via npm canonicalizeJCS`)

// --- 3. Reference model -----------------------------------------------------

let matched = 0
for (const c of vectors.cases) {
  const got = evaluateReference(c.group, c.input)
  const same =
    got.chain_verdict === c.expected.chain_verdict &&
    got.verdict === c.expected.verdict &&
    got.reason === c.expected.reason
  if (!same) {
    fail(
      `${c.id}: reference gave ${got.chain_verdict}/${got.verdict}/${got.reason}, ` +
        `expected ${c.expected.chain_verdict}/${c.expected.verdict}/${c.expected.reason}`,
    )
  } else matched++
}
console.log(`  ok   reference model: ${matched}/${vectors.cases.length} matched`)

// --- 4. Negative controls ---------------------------------------------------

for (const control of vectors.policies.negative_controls) {
  if (CONTROL_FOR_GROUP[control.group] !== control.id) {
    fail(`${control.id}: harness binds group ${control.group} to ${CONTROL_FOR_GROUP[control.group]}`)
    continue
  }
  const inGroup = vectors.cases.filter(c => c.group === control.group)
  const observed: string[] = []
  for (const c of inGroup) {
    const got = evaluateControl(control.group, c.input)
    const same =
      got.chain_verdict === c.expected.chain_verdict &&
      got.verdict === c.expected.verdict &&
      got.reason === c.expected.reason
    if (!same) observed.push(c.id)
  }
  const declared = [...control.declared_fail_set].sort()
  if (JSON.stringify(declared) !== JSON.stringify([...observed].sort())) {
    fail(`${control.id}: declared fail set ${JSON.stringify(declared)}, observed ${JSON.stringify(observed.sort())}`)
  } else {
    console.log(
      `  ok   ${control.id}: ran ${inGroup.length} cases in group ${control.group}, failed exactly ${declared.length} declared (${declared.join(', ')})`,
    )
  }
  if (declared.length === 0) fail(`${control.id}: a negative control with an empty declared fail set is not a control`)
}

// --- 5. Input immutability --------------------------------------------------

let unchanged = 0
for (const c of vectors.cases) {
  const after = digestOf(c.input)
  if (after !== before.get(c.id)) fail(`${c.id}: an evaluator mutated the input it was given`)
  else unchanged++
}
console.log(`  ok   input immutability: ${unchanged}/${vectors.cases.length} inputs unchanged after every evaluator ran`)

if (failures.length > 0) {
  console.error(`\nlifecycle-evidence-and-record TypeScript: FAILED, ${failures.length} problem(s)`)
  process.exit(1)
}
console.log(`\nlifecycle-evidence-and-record TypeScript: ${vectors.cases.length}/${vectors.cases.length} passed`)
